### PR TITLE
Streamline ManageSurfaceHeatBalance - Part 1

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
Initial work on cleaning up `CalcHeatBalanceInsideSurf`.

- [x] Replace `any_eq` with simple bools for which algorithms are in use
- [x] Move out things that should only happen once, like building the heat transfer surface list, etc.
- [x] Move out things that should only happen at the end of the timestep (not repeatedly during iterations), like computing zone window heat gains, etc.
- [x] Build vectors with surface lists to drive loops in CalcHeatBalanceInsideSurf to reduce repeated checking if a surface is a heat transfer surface, or a window, etc.
- [x] Add unit tests (except none yet for CalcHeatBalanceInsideSurf, because more work is planned there)

*Stopping here for now. Will take up the remaining items and more in a new PR building on this one.*
- [ ] Move Kiva surface temp set and reset out of iteration loop - tried but got diffs - likely due to the every-30-iteration call to `InitInteriorConvectionCoeffs`
- [ ] Streamline calcs
- [ ] ??

### Diff Summary
WCE* small eso diffs in ROOM 102 WINDOW:Surface Inside Face Convection Heat Gain Rate. The new values appear to be better than before - they match h*DeltaT.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by nrel-bot-2** [comment]
@mjwitte it has been 22 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@mjwitte it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@mjwitte it has been 8 days since this pull request was last updated.

**Comment by mjwitte** [comment]
@Myoldmopar Code is ready for initial review, but still need to add unit tests.

File: src/EnergyPlus/DataSurfaces.cc [hunk 1]
```diff
@@ -498,6 +498,12 @@ namespace DataSurfaces {
     Array1D<Real64> WinGapConvHtFlowRepEnergy;     // Energy of WinGapConvHtFlowRep [J]
     Array1D<Real64> WinHeatTransferRepEnergy;      // Energy of WinHeatTransfer [J]
 
+    std::vector<int> AllHTSurfaceList;  // List of all heat transfer surfaces
+    std::vector<int> AllIZSurfaceList;  // List of all interzone heat transfer surfaces
+
+    bool AnyHeatBalanceInsideSourceTerm(false);  // True if any SurfaceProperty:HeatBalanceSourceTerm inside face used
+    bool AnyHeatBalanceOutsideSourceTerm(false); // True if any SurfaceProperty:HeatBalanceSourceTerm outside face used
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataSurfaces.cc:507]: These flags and specialized vectors will surely be helpful.

File: src/EnergyPlus/DataSurfaces.hh [hunk 2]
```diff
@@ -1093,6 +1099,8 @@ namespace DataSurfaces {
         int WindowModelType;           // if set to WindowBSDFModel, then uses BSDF methods
         BSDFWindowDescript ComplexFen; // Data for complex fenestration, see DataBSDFWindow.cc for declaration
 
+        int TDDPipeNum; // Tubular daylighting device pipe number for TDD domes and diffusers
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataSurfaces.hh:1104]: Yes, skipping looking this up each time will be an easy win for sure.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 3]
```diff
@@ -184,40 +184,20 @@ namespace HeatBalanceSurfaceManager {
...
     // PUBLIC  CalcHeatBalanceInsideSurf   ! radiant systems need access to them in order to simulate
 
-    // Record Keeping/Utility Routines for Module
-
-    // Reporting routines for module
-
-    // MODULE SUBROUTINES:
-    //*************************************************************************
-
-    // Functions
```
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:220]: I like all the cleanup.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 4]
```diff
@@ -5826,8 +5781,55 @@ namespace HeatBalanceSurfaceManager {
 
     void CalcHeatBalanceInsideSurf(Optional_int_const ZoneToResimulate) // if passed in, then only calculate surfaces that have this zone
     {
+        // Pass correct list of surfaces to CalcHeatBalanceInsideSurf2
+        bool const PartialResimulate(present(ZoneToResimulate));
 
-        // SUBROUTINE INFORMATION:
+        if (!PartialResimulate) {
+            for (int zoneNum = 1; zoneNum <= NumOfZones; ++zoneNum) {
+                ZoneWinHeatGain = 0.0;
```
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Is this assigning zero to each vector item each time this loop occurs?

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 5]
```diff
@@ -5856,59 +5858,17 @@ namespace HeatBalanceSurfaceManager {
...
-        using DataMoistureBalance::HMassConvInFD;
-        using DataMoistureBalance::HSkyFD;
-        using DataMoistureBalance::RhoVaporAirIn;
-        using DataMoistureBalance::RhoVaporAirOut;
-        using DataMoistureBalance::RhoVaporSurfIn;
-        using DataMoistureBalance::TempOutsideAirFD;
-        using DataMoistureBalanceEMPD::HeatFluxLatent;
-        using DataMoistureBalanceEMPD::RVSurface;
-        using DataMoistureBalanceEMPD::RVSurfaceOld;
-        using DataMoistureBalanceEMPD::RVSurfLayer;
```
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:5887]: I'm liking all these usings going away :+1:

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 6]
```diff
@@ -6099,81 +6013,66 @@ namespace HeatBalanceSurfaceManager {
...
         }
 
-        bool const useCondFDHTalg(any_eq(HeatTransferAlgosUsed, UseCondFD));
         Converged = false;
         while (!Converged) { // Start of main inside heat balance DO loop...
 
             TempInsOld = TempSurfIn; // Keep track of last iteration's temperature values
 
-            if (any_eq(HeatTransferAlgosUsed, HeatTransferModel_Kiva)) {
+            if (DataHeatBalance::AnyKiva) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:6069]: This should be a really good gain.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 7]
```diff
@@ -6495,30 +6383,27 @@ namespace HeatBalanceSurfaceManager {
...
-                        if (construct.SourceSinkPresent && calcHeatBalanceInsideSurfFirstTime)
-                            ShowSevereError("Windows are not allowed to have embedded sources/sinks");
-
                         if (SurfaceWindow(SurfNum).OriginalClass == SurfaceClass_TDD_Diffuser) { // Tubular daylighting device
                             // Lookup up the TDD:DOME object
-                            PipeNum = FindTDDPipe(SurfNum);
-                            SurfNum2 = TDDPipe(PipeNum).Dome;
-                            Ueff = 1.0 / TDDPipe(PipeNum).Reff;
+                            int pipeNum = SurfaceWindow(SurfNum).TDDPipeNum;
+                            int domeNum = DataDaylightingDevices::TDDPipe(pipeNum).Dome;
```
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc]: The compiler is probably already doing it, but I'd assign `const` where possible to be clear this is just assigned.

**Review by Myoldmopar — COMMENTED** [review]
This is going to be really great.  I made some comments and had one question, but overall this is in good shape.

**Comment by Myoldmopar** [comment]
I see there's a few percent gain on the current test suite, but do you have annual results?  I'd expect them to be even higher.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc (hunk 4 above)
Comment (mjwitte) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Yes. Suggestions?

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc (hunk 7 above)
Comment (mjwitte) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: OK - will do.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc (hunk 4 above)
Comment (Myoldmopar) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Well, I mean it looks like ZoneWinHeatGain is an array allocated to the number of zones.  Just like the other arrays here.  Assigning a zero value to the ZoneWinHeatGain array here is essentially doing an internal loop over all the zones and assigning every ZoneWinHeatGain index equal to zero.  

I would either change this line to `ZoneWinHeatGain(zoneNum) = 0.0;`, like this:

```
for (int zoneNum = 1; zoneNum <= NumOfZones; ++zoneNum) {
  ZoneWinHeatGain(zoneNum) = 0.0;   // THIS IS NOW INDEXED
  ZoneWinHeatGainRep(zoneNum) = 0.0;
  ZoneWinHeatGainRepEnergy(zoneNum) = 0.0;
  ZoneWinHeatLossRep(zoneNum) = 0.0;
  ZoneWinHeatLossRepEnergy(zoneNum) = 0.0;
}
```

... or just take all these arrays out of the loop here and assign them all to zero at once:

```
ZoneWinHeatGain = 0.0;
ZoneWinHeatGainRep = 0.0;
ZoneWinHeatGainRepEnergy = 0.0;
ZoneWinHeatLossRep = 0.0;
ZoneWinHeatLossRepEnergy = 0.0;
```

In that case, the loop would occur for each, but you would rely on some possibly optimized operations inside of each array and avoid the array lookup operations.  It would be interesting to do a test to see which is faster.

OR

You tell me why I missed the point of this operation and it's actually correct as-is. 😄

Comment (mjwitte) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc]: Got it. These can all be simply `ZoneWinHeatGain* = 0.0;` and drop the for loop.  And this screams for a window surface lists. Adding that as well.

**Comment by mjwitte** [comment]
@Myoldmopar I've addressed your comments and went a step further with separate surface lists for windows and non-windows.

Annual run comparison of 10x_incr_100zones100surfaces.idf (non-scientific, single-run stats on laptop, and yes, this file has 10,200 surfaces, but no windows):
Develop 9.2.0-26151c5477 24 min 54 sec
This branch 9.2.0-39ff270514 22 min 50 sec

So, would you like me to stop on this branch and start a new one for further work (just to keep things in reviewable chunks), or keep going here?

**Comment by Myoldmopar** [comment]
I'd say leave it as-is for this branch.  This is a really nice cleanup.  I see the WCE diffs have been present for quite a while, have you already investigated those or should I dive in?

**Comment by mjwitte** [comment]
Comment lurking at the top

> Diff Summary
WCE* small eso diffs in ROOM 102 WINDOW:Surface Inside Face Convection Heat Gain Rate. The new values appear to be better than before - they match h*DeltaT.

Would be prudent for you to look at one of these to see if you agree.

**Comment by Myoldmopar** [comment]
OK, built this locally, pulled in latest develop and compared to develop. I reproduced the small diffs in the WCE files.  I calculated the convection rate manually in both files and found that while it may be closer, it's deep in the decimal points and insignificant.  This is good to go in.  I'm going to merge.
